Template Variables on contained type

moxcodes opened the pull request:
## Proposed changes

- Tensor is templated on type, so Variables can easily be templated on the type
  to be stored in the Tensor.
- Necessary for usage of extensions such as ComplexDataVector
- Update tests to be a single test case and to use random variables where
  possible

New data structures which take advantage of this templating should add tests to the Variables test case, as well as extending the Tensor static_asserts. I will do this for ComplexDataVector, and include additional documentation in the datavectorimpl PR explaining these extra implications.  

Note: the type inference from the tags now requires there to be at least 1 tag for the variables class. I understand it was deliberately permitted to have a tagless variables type, but the only thing that broke when I added that `static_assert` was the empty variables test case. So, unless there is a compelling reason to add extra syntax (probably SFINAE) around an empty variables, I think it seems cleaner to just demand non-empty variables.

### Logical order of the commits
- Add utilitiy for making signaling NaNs
- Template Variables on contained type
- Convert test of variables construct and assign
- convert variables move test to template
- convert variables math test for template
- convert variables prefix move semantics test
- convert template variables prefix arithmetic test
- convert variables serialization test to template
- convert variables test assign subset to template
- convert variables slice utilities test to template
- convert test for adding slice to variables
- convert assertion variable tests to template

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsdeppe reviewed (changes requested):
This is inconsistent with the type of the tag. You should get the underlying type from the Tags not having it passed in.

moxcodes commented:
@nilsdeppe good point, that also should fix the forward declaration frustration, as it would no longer need to be templated on the storage type.

moxcodes commented:
This has been fixed to appropriately use the type from the tags

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -88,17 +94,30 @@ class Variables;
+
   using tags_list = tmpl::list<Tags...>;
+  static_assert(sizeof...(Tags) > 0,
+                "You must provide at least one tag to the variables"
```
`variables` -> `Variables ` (capital and space after)

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -88,17 +94,30 @@ class Variables;
+      tmpl2::flat_all_v<cpp17::is_same_v<
+          typename Tags::type::type,
+          typename tmpl::at<tags_list, tmpl::size_t<0>>::type::type>...>,
+      "All tensors stored in a single Variables must"
```
space after `must`

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -88,17 +94,30 @@ class Variables;
+  static_assert(
+      tmpl2::flat_all_v<cpp17::is_same_v<
+          typename Tags::type::type,
+          typename tmpl::at<tags_list, tmpl::size_t<0>>::type::type>...>,
```
You can use `tmpl::front<tags_list>` here :)

nilsdeppe replied:
ping

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -223,17 +242,15 @@ class Variables<tmpl::list<Tags...>> {
       const Variables<tmpl::list<SubsetOfTags...>>& vars) noexcept {
-    (void)std::initializer_list<char>{
-        (get<SubsetOfTags>(*this) = get<SubsetOfTags>(vars), '0')...};
+    expand_pack((get<SubsetOfTags>(*this) = get<SubsetOfTags>(vars))...);
```
These are no longer the same. (`EXPAND_PACK_LEFT_TO_RIGHT` would be the same)

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -223,17 +242,15 @@ class Variables<tmpl::list<Tags...>> {
       const tuples::TaggedTuple<SubsetOfTags...>& vars) noexcept {
-    (void)std::initializer_list<char>{
-        (get<SubsetOfTags>(*this) = get<SubsetOfTags>(vars), '0')...};
+    expand_pack((get<SubsetOfTags>(*this) = get<SubsetOfTags>(vars))...);
```
These are no longer the same.

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -426,13 +439,15 @@ void Variables<tmpl::list<Tags...>>::initialize(
-    std::fill(variable_data_impl_.get(), variable_data_impl_.get() + size_,
-              std::numeric_limits<double>::signaling_NaN());
+    if (cpp17::is_arithmetic_v<value_type>) {
+      std::fill(variable_data_impl_.get(), variable_data_impl_.get() + size_,
```
Does this actually compile if the tensor holds `Complex`? It would if we had `if constexpr` but as is I'm not so sure. Would something like `make_with_value<value_type>(0., NaN);` work? I think for complex numbers we still probably want them NaN-initialized in debug mode

moxcodes replied:
It does compile, but it also skips filling with NaNs - I can probably do something with `make_with_value` to get it to work.

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -675,13 +690,16 @@ void Variables<tmpl::list<Tags...>>::add_reference_variable_data(
                                              << " and " << rhs.size());
-  double* const lhs_data = lhs.data();
-  const double* const rhs_data = rhs.data();
+  typename Variables<tmpl::list<Tags...>>::value_type* const lhs_data =
```
maybe `using value_type = ...;` to shorten this a bit? (same below)

nilsdeppe commented on src/DataStructures/VariablesHelpers.hpp:
```diff
@@ -59,8 +59,9 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
   constexpr const size_t number_of_independent_components =
       Variables<TagsList>::number_of_independent_components;
   Variables<TagsList> interface_vars(interface_grid_points);
-  const double* vars_data = vars.data();
```
A type alias here would help too

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -3,14 +3,13 @@
 
 #include "tests/Unit/TestingFramework.hpp"
 
-#include <array>
```
Is there no way to present the test changes in multiple commits? This will be very difficult to review and convince ourselves it is still correct.

moxcodes replied:
Sure, I can do that. Which of the following is preferable for reviewing?
1. two commits, one for direct templatization, one for changing hard-coded variables to more widely applicable randomly generate values
2. several commits, one for each test case converted.

nilsdeppe replied:
Good question. I think 1. would be fine, though 2. sounds even easier to review. I'll let you choose and I'll deal with the fallout :P

moxcodes committed:
Convert Variables tests to new template form

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -88,17 +93,30 @@ class Variables;
-                    blaze::DynamicVector<double, transpose_flag>>;
-
   /// A typelist of the Tags whose variables are held
+
```
Please don't add this line

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -88,17 +93,30 @@ class Variables;
+      "All tensors stored in a single Variables must"
+      "have the same internal storage type.");
+
+  using vector_type = typename tmpl::at<tags_list, tmpl::size_t<0>>::type::type;
```
`tmpl::front`

moxcodes replied:
some of these are already in the fixup. I'll squash so it's less tricky to see :)

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+      typename tt::get_fundamental_type_t<typename VectorType::value_type>>
+      dist{-100, 100};
+
+  size_t number_of_grid_points = 1;
```
We've been preferring 5 grid points. Is that compatible with this test?

moxcodes replied:
It makes the string-generation test below less succinct. My suspicion is that is why 1 was used here in the previous version as well.

nilsdeppe replied:
Okay, maybe we should have a 2 point vector for the stream test, and 5 for the others then :) Just to make sure stream is working correctly with multiple grid points.

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
-  CHECK(v != v2);
-  v2 = v;
-  CHECK(v2 == v);
+  vector_in_filled_variables =
```
`vector` -> `tensor`? I find that you are using `vector` for both PointerVector-ish meaning and rank 1 tensor confusing.

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
 
-  const auto& kvector_in_v = get<VariablesTestTags_detail::vector>(v);
-  CHECK(kvector_in_v.get(0)[0] == -4.0);
+  auto set_value = make_with_random_values<typename VectorType::value_type>(
```
Why `set_value`? Maybe this is a good time to give things better names than `another_another_value` (which I think is what `set_value` should be called)

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+  CHECK(initial_fill_value == vector_in_filled_variables.get(0)[0]);
 
-  tnsr::I<DataVector, 3, Frame::Grid> another_vector(1_st, -5.0);
+  auto another_value = make_with_random_values<typename VectorType::value_type>(
```
Maybe `value_for_reference_assignment`?

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+      make_not_null(&gen), make_not_null(&dist));
 
-  CHECK(-5.0 == another_vector.get(0)[0]);
+  tnsr::I<VectorType, 3, Frame::Grid> another_vector{number_of_grid_points,
```
`tensor_for_reference_assignment`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+  CHECK(number_of_grid_points == filled_variables.number_of_grid_points());
+  CHECK(number_of_grid_points * 5 == filled_variables.size());
+
+  auto& vector_in_filled_variables =
```
`tensor_in_filled_variables`

nilsdeppe replied:
Please add a comment `// Test that assigning to a tensor pointing into a Variables correctly sets the Variables`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+  another_filled_variables = filled_variables;
+  CHECK(another_filled_variables == filled_variables);
+
+  Variables<tmpl::list<VariablesTestTags_detail::vector_tag<VectorType>,
```
Please add a comment: `// Test default constructed variables is allowed`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+  default_variables = another_filled_variables;
+  CHECK(another_filled_variables == default_variables);
 
   const std::string expected_output =
```
Then add a comment here `// Test stream operator`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+      dist{-100, 100};
+
+  size_t number_of_grid_points = 1;
+  auto initial_fill_value =
```
Please add a comment `// Test constructed and member function initialized variables are identical`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+  CHECK(set_value == vector_in_filled_variables.get(1)[0]);
+  CHECK(set_value == vector_in_filled_variables.get(2)[0]);
+
+  const auto& kvector_in_filled_variables =
```
Please add a comment `// Test const vector in variables points to right values`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,108 @@ static_assert(
+      get<VariablesTestTags_detail::vector_tag<VectorType>>(filled_variables);
+  CHECK(kvector_in_filled_variables.get(0)[0] == set_value);
+
+  Variables<tmpl::list<VariablesTestTags_detail::vector_tag<VectorType>,
```
Please add a comment `// Test equivalence and inequivalence operators`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -601,6 +634,28 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Variables.add_slice_to_data",
         get<VariablesTestTags_detail::vector>(vars));
 }
 
+SPECTRE_TEST_CASE("Unit.DataStructures.Variables", "[DataStructures][Unit]"){
```
Formatting `){` and 2 space indentation. Please try using `git clang-format -f` before committing. It will format only the lines you touched :)

moxcodes replied:
sorry, this one's rough because clang-format removes all indentation, getting confused from the `SECTION`s, so had to resort to (fallible) manual formatting .

nilsdeppe replied:
Ah, getting. I've encountered clang-format's dislike of `SECTION` too, but don't understand it. Understood what the struggle was :)

nilsdeppe replied:
What I've done when using sections is `// clang-format off` and then later `// clang-format on`. Sometimes clang-format gets things really wrong... (I'm not asking you to change anything here, just informing you so you can do what you prefer :) It looks like in the next commit clang-format smartens up anyway :) )

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -229,30 +229,51 @@ void test_variables_construction_and_access() noexcept {
+  UniformCustomDistribution<
+      typename tt::get_fundamental_type_t<typename VectorType::value_type>>
+      dist{-100, 100};
+  UniformCustomDistribution<size_t> sdist{2, 20};
```
Let's do 5 -> 20 :) We've had 2 points let us sneak problems into the code in the past

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -229,30 +229,51 @@ void test_variables_construction_and_access() noexcept {
+      dist{-100, 100};
+  UniformCustomDistribution<size_t> sdist{2, 20};
+
+  size_t number_of_grid_points1 = sdist(gen),
```
`const`? We prefer specifying the type on both lines (something I disagreed with for a long time but have come to appreciate a lot since reading other people's code more)

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -229,30 +229,51 @@ void test_variables_construction_and_access() noexcept {
+
+  size_t number_of_grid_points1 = sdist(gen),
+         number_of_grid_points2 = sdist(gen);
+  std::vector<typename VectorType::value_type> initial_fill_values{2};
```
You want a single `std::vector` with the value `2` and then fill it randomly?

moxcodes replied:
should be the size with a single value specified. Here I'm trying to aggregate the random generations of similarly used values so that `make_with_random_values` doesn't have to appear so frequently.

nilsdeppe replied:
You are creating a `std::vector` with one element whose value is `2`. Is this what you want or do you want a vector with 2 elements? I think if you use the `make_with_random_value` you can make this const and avoid the confusing mess that is constructing `std::vector`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -229,30 +229,51 @@ void test_variables_construction_and_access() noexcept {
+                          make_not_null(&gen), make_not_null(&dist));
+
+  Variables<tmpl::list<VariablesTestTags_detail::vector_tag<VectorType>>>
+      target{number_of_grid_points1, initial_fill_values[0]};
```
maybe prefer `initial_fill_values.at(0)` to get bounds checking?

nilsdeppe replied:
We normally use `target` to mean the `target points of a projection operation`. I'm not sure what it's supposed to mean in this context.

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -229,30 +229,51 @@ void test_variables_construction_and_access() noexcept {
-         Variables<tmpl::list<VariablesTestTags_detail::vector>>{2, -3.0}));
-  CHECK(&get<VariablesTestTags_detail::vector>(x)[0][0] == x.data());
+  CHECK(
+      (target ==  // NOLINT
```
We no longer need the extra parens

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<
+      typename tt::get_fundamental_type_t<typename VectorType::value_type>>
+      dist{-100, 100};
```
Is including zero possibly going to cause division problems?

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  UniformCustomDistribution<
+      typename tt::get_fundamental_type_t<typename VectorType::value_type>>
+      dist{-100, 100};
+  UniformCustomDistribution<size_t> sdist{2, 20};
```
5 -> 20

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
-  check_vectors(test_variable_type{1, 19.0}, test_assignment);
-  test_assignment -= test_variable_type{1, 3} * 2.0;
-  check_vectors(test_variable_type{1, 13.0}, test_assignment);
+  size_t s = sdist(gen);  // number of grid points
```
const? In general, can more things in these tests become const?

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  size_t s = sdist(gen);  // number of grid points
+  auto vars_val = make_with_random_values<typename VectorType::value_type>(
+      make_not_null(&gen), make_not_null(&dist));
+  std::vector<typename VectorType::value_type> rvs{3};
```
`rvs`?

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
-  test_assignment -= test_variable_type{1, 3} * 2.0;
-  check_vectors(test_variable_type{1, 13.0}, test_assignment);
+  size_t s = sdist(gen);  // number of grid points
+  auto vars_val = make_with_random_values<typename VectorType::value_type>(
```
`vars_val`?

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  auto vars_val = make_with_random_values<typename VectorType::value_type>(
+      make_not_null(&gen), make_not_null(&dist));
+  std::vector<typename VectorType::value_type> rvs{3};
+  fill_with_random_values(
```
Why not just use `make_with_random_values` everywhere and make things `const`?

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
-  check_vectors(test_variable_type{1, 19.0}, test_assignment);
-  test_assignment -= test_variable_type{1, 3} * 2.0;
-  check_vectors(test_variable_type{1, 13.0}, test_assignment);
+  size_t s = sdist(gen);  // number of grid points
```
`s` -> `number_of_grid_points`

moxcodes replied:
for this math test case, I defaulted back to short but ambiguous variable names because clearer ones make the math tests with many operands span several lines. I've tried to clarify with comments near the declaration site. Do you still prefer longer, clearer variable names if it multiplies the line length of this test by 2-3?

nilsdeppe replied:
Yes, `num_points` might be a better middle ground. We generally prefer names that make comments unnecessary because if you change a name of a variable the compiler will complain if you don't do it consistently, but it won't ask you to update the comment :)

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  check_vectors(test_variable_type{s, 4. * vars_val},
+                vars + vars + vars + vars);
+  check_vectors(test_variable_type{s, 3. * vars_val}, vars + (vars + vars));
+  // clang-tidy: point sides of overloaded operator are equivalent
```
`point sides`? Do you mean `both sides`?

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
-                  "[DataStructures][Unit]") {
+template <typename VectorType>
+void test_variables_math() noexcept {
   using test_variable_type =
```
Should be CamelCase (it's used as a type, not a metavariable)

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
-  test_datavector_math /= dv;
-  check_components(test_datavector_math, {2., 4., 6., 8.});
+
+  VectorType vec{4};
```
Could you add a comment about what the below tests are doing compared to the above tests? Like what I had suggested on a previous.

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  check_vectors(test_variable_type{s, vars_val}, vars - vars + vars);  // NOLINT
+  check_vectors(test_variable_type{s, vars_val}, vars + vars - vars);
+
+  test_variable_type test_assignment(vars * rvs[0]);
```
Please add a comment `// Test math assignment operators +=, -=, *=, /=`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -300,66 +300,104 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+      make_not_null(&gen), make_not_null(&dist));
 
-  test_variable_type test_assignment2(1, 0.0);
+  test_variable_type vars{s, vars_val};
```
Please add a comment `// Test math +,-,*,/`

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -400,20 +400,35 @@ void test_variables_math() noexcept {
+  UniformCustomDistribution<
+      typename tt::get_fundamental_type_t<typename VectorType::value_type>>
+      dist{-100, 100};
+  UniformCustomDistribution<size_t> sdist{2, 20};
```
5 -> 20

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -400,20 +400,35 @@ void test_variables_math() noexcept {
+
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<
+      typename tt::get_fundamental_type_t<typename VectorType::value_type>>
```
In general, the whole point of `_t` is so you don't need to add `typename` in front. Please remove the extra `typename` everywhere you're using it.

nilsdeppe reviewed (changes requested):
I'm going to let you propagate the requested changes throughout the code since a lot of them will be recurring in almost every test commit. Please squash the fixups into the appropriate commit of the ones you currently have :)

moxcodes commented:
Changes squashed in.
Cheers!

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
-  CHECK(5 == v.size());
+namespace {
+std::string repeat_string(const std::string& str, size_t repeats) noexcept {
+  if (repeats <= 0) {
```
Can't be less than zero, `size_t` is unsigned (also, repeat should be const). Can't this be done with a for loop? Something like:
```cpp
MakeString t;
for (size_t i = 0; i < repeats; ++i) {
  t << str;
}
return t;
```
(MakeString implicitly converts to a string)

moxcodes commented on src/DataStructures/Variables.hpp:
```diff
@@ -88,17 +93,29 @@ class Variables;
+
+  static_assert(
+      tmpl2::flat_all_v<
+          cpp17::is_same_v<typename Tags::type::type,
```
static assert also that the type has a value type and give a good error message when it doesn't.

moxcodes commented on src/DataStructures/Variables.hpp:
```diff
@@ -426,13 +439,14 @@ void Variables<tmpl::list<Tags...>>::initialize(
+               number_of_independent_components * sizeof(value_type))));
     number_of_grid_points_ = number_of_grid_points;
 #if defined(SPECTRE_DEBUG) || defined(SPECTRE_NAN_INIT)
     std::fill(variable_data_impl_.get(), variable_data_impl_.get() + size_,
```
create a user-defined signalling-NaN function that works well for double and std::complex.

nilsdeppe replied:
wouldn't just `make_with_value<value_type>(...)` filled with NaN work if passed to `std::fill`? That might be easier than adding another function, if it works :)

moxcodes commented on src/DataStructures/Variables.hpp:
```diff
@@ -693,28 +709,33 @@ Variables<tmpl::list<Tags...>>& operator*=(Variables<tmpl::list<Tags...>>& lhs,
                                              << " and " << rhs.size());
-  double* const lhs_data = lhs.data();
-  const double* const rhs_data = rhs.data();
+  typename Variables<tmpl::list<Tags...>>::value_type* const lhs_data =
```
do `using value_type` here too

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
+  if (repeats <= 0) {
+    return "";
+  }
+  return str + repeat_string(str, repeats - 1);
```
prefer loop

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
-  CHECK(-5.0 == v.data()[0]);  // NOLINT
-  CHECK(-5.0 == vector_in_v.get(0)[0]);
+  CHECK(value_for_reference_assignment ==
+        tensor_for_reference_assignment.get(0)[0]);
```
`get(0)` should be `get<0>`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
-  CHECK(-3.0 == v.data()[0]);  // NOLINT
-  CHECK(-3.0 == vector_in_v.get(0)[0]);
+  CHECK(initial_fill_value == filled_variables.data()[0]);  // NOLINT
+  CHECK(initial_fill_value == tensor_in_filled_variables.get(0)[0]);
```
`get(0)` should be `get<0>`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
   // clang-tidy: do not use pointer arithmetic
-  CHECK(-3.0 == v.data()[0]);  // NOLINT
-  CHECK(-3.0 == vector_in_v.get(0)[0]);
+  CHECK(initial_fill_value == filled_variables.data()[0]);  // NOLINT
```
this test should be grouped with the block ending on line 140.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
-  CHECK(-4.0 == vector_in_v.get(2)[0]);
+  CHECK(value_for_reference_assignment ==
+        filled_variables.data()[0]);  // NOLINT
+  CHECK(value_for_reference_assignment == tensor_in_filled_variables.get(0)[0]);
```
`get<0>`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
+  CHECK(value_for_tensor_constructor == filled_variables.data()[0]);  // NOLINT
+  CHECK(value_for_tensor_constructor == filled_variables.data()[1]);  // NOLINT
+  CHECK(value_for_tensor_constructor == filled_variables.data()[2]);  // NOLINT
+  CHECK(value_for_tensor_constructor == tensor_in_filled_variables.get(0)[0]);
```
`get<>`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
+  // Test const vector in variables points to right values
+  const auto& ktensor_in_filled_variables =
+      get<VariablesTestTags_detail::tensor_tag<VectorType>>(filled_variables);
+  CHECK(ktensor_in_filled_variables.get(0)[0] == value_for_tensor_constructor);
```
`get<>`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
+  CHECK(ktensor_in_filled_variables.get(0)[0] == value_for_tensor_constructor);
+
+  // Test equivalence and inequivalence operators
+  Variables<tmpl::list<VariablesTestTags_detail::tensor_tag<VectorType>,
```
use `decltype` to avoid repeating the long Variables type. (and below)

nilsdeppe replied:
or a type alias? I'm fine with either :)

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
+                       VariablesTestTags_detail::scalar_tag<VectorType>,
+                       VariablesTestTags_detail::scalar2_tag<VectorType>>>
+      another_filled_variables{number_of_grid_points, initial_fill_value};
+  CHECK(filled_variables != another_filled_variables);
```
add a comment `//handle case of same value rolled twice on the random number generators`
`CHECK((filled_variables == another_filled_variables)
  == (initial_fill_value == value_for_tensor_constructor))`

wthrowe replied:
I think you meant to delete this line.  It doesn't seem to make sense alongside the new one you added.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,79 +103,136 @@ static_assert(
-  CHECK(get_output(empty_vars) == "Variables is empty!");
+      "tensor_tag:\n"
+      "T(0)=(" +
+      repeat_string(get_output(value_for_tensor_constructor) + ",",
```
above this `output_string = repeat_string(get_output...` to avoid this so many times, or even a lambda to make the precomputation lines concise.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -166,34 +241,19 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Variables", "[DataStructures][Unit]") {
 #endif  // defined(__clang__) && __clang_major__ > 6
 #endif  // ! __APPLE__
-  v = v;
+  filled_variables = filled_variables;  //NOLINT
```
space after `//` (check clang-format on this file)

moxcodes commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -180,7 +180,8 @@ template <typename T, typename UniformRandomBitGenerator,
     const gsl::not_null<UniformRandomBitGenerator*> generator,
     const gsl::not_null<RandomNumberDistribution*> distribution) noexcept {
-  T result{std::numeric_limits<tt::get_fundamental_type_t<T>>::signaling_NaN()};
+  T result
```
make this change only in one commit. And in whichever commit the real change is, make sure that it calls `fill` for container types, and just constructs with `NaN` for fundamental types.

moxcodes reviewed (commented):
Group code review with @markscheel and @fmahebert

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -788,21 +788,33 @@ void test_variables_assign_subset() noexcept {
+template <typename VectorType>
+void test_variables_slice() noexcept {
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<size_t> sdist{5, 20};
```
We should limit this one to `10` since we currently allow at most 12 grid points per dimension :)

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -858,44 +858,82 @@ void test_variables_slice() noexcept {
+  // The slice0_vals should have been added to the second half of each of the
+  // three vectors in the tensor. The slice1_vals should have been added to
+  // entries 2 and 6 in each vector.
+  auto expected = Tensor{
```
`const auto expected` or `const Tensor expected{`

nilsdeppe reviewed (commented):
I'm happy for you to squash into the current commits. Once you've fixed your group review comments and mine we can ping the other Cornell folks to look at this if they want to.

Thanks for cleaning up the mess that was the variables tests :)

moxcodes commented on src/DataStructures/Variables.hpp:
```diff
@@ -172,7 +189,7 @@ class Variables<tmpl::list<Tags...>> {
   /// the constructor with the same arguments.
   void initialize(size_t number_of_grid_points) noexcept;
-  void initialize(size_t number_of_grid_points, double value) noexcept;
+  void initialize(size_t number_of_grid_points, value_type value) noexcept;
```
add comment that indicates this should be updated if we ever use a Variables which has `value_type` larger than ~2 doubles in size.

moxcodes commented on src/DataStructures/Variables.hpp:
```diff
@@ -88,17 +93,29 @@ class Variables;
+                           typename tmpl::front<tags_list>::type::type>...>,
+      "All tensors stored in a single Variables must "
+      "have the same internal storage type.");
+
```
add a `static_assert` to ensure that `value_type` is a fundamental type or a std::complex, with a message that indicates the reasoning.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -328,66 +328,120 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
-  test_assignment -= test_variable_type{1, 3} * 2.0;
-  check_vectors(test_variable_type{1, 13.0}, test_assignment);
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<tt::get_fundamental_type_t<value_type>> dist{1,
```
prefer `1.0`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -328,66 +328,120 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  // Test math +, -, *, /
+  const TestVariablesType vars{num_points, value_in_variables};
+  check_vectors(
+      TestVariablesType{num_points, rand_vals.at(0) * value_in_variables},
```
now that this is array, `at(...)` -> `[...]`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -328,66 +328,120 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+
+  // Test math +, -, *, /
+  const TestVariablesType vars{num_points, value_in_variables};
+  check_vectors(
```
change the name to `check_variables`, but in a final, separate commit (so that the other tests compile in the intermediate commits)

nilsdeppe replied:
I'd be fine with the final product that we merge being one big commit, just as an FYI :)

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -328,66 +328,120 @@ void check_vectors(const Variables<T1>& t1, const Variables<T2>& t2) {
+  TestVariablesType test_vector_math(4, rand_vals.at(0));
+  test_vector_math *= test_vector;
+  VectorType vec_expect = rand_vals.at(0) * test_vector;
+  check_components(test_vector_math, {vec_expect.at(0), vec_expect.at(1),
```
just use `vec_expect` instead of expanding and using the initializer list.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -444,24 +444,45 @@ void test_variables_math() noexcept {
+                           VariablesTestTags_detail::scalar_tag<VectorType>,
+                           VariablesTestTags_detail::scalar2_tag<VectorType>>>;
+  using PrefixVariablesType = Variables<
+      tmpl::list<VariablesTestTags_detail::PrefixTag0<
```
use ~`add_tags_to`~ `wrap_tags_in`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -444,24 +444,45 @@ void test_variables_math() noexcept {
 
-  const auto check_variables = [](const auto& lvars_0, const auto& lvars_1) {
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<tt::get_fundamental_type_t<value_type>> dist{-100,
```
decimal point (and all other places the distribution is made)

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -444,24 +444,45 @@ void test_variables_math() noexcept {
+  VariablesType move_constructed_vars{
+    std::move(prefix_vars_move_construct_from)};
+
+  const auto check_variables = [](const auto& lvars_0,
```
should be able to just use `check_vectors` and replace the lambda.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -470,15 +491,23 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Variables.PrefixSemantics",
-  check_variables(vars_0, variable_type(4, 4.0));
-  vars_0 = std::move(prefix_vars_3);
-  check_variables(vars_0, variable_type(4, 2.0));
+  check_variables(vars_to,
```
all of these helper check functions should just call a CHECK_VARIABLES_APPROX in `TestingFramework.hpp`. @nilsdeppe @wthrowe is there a good reason this shouldn't be done?

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -510,69 +510,128 @@ void test_variables_prefix_semantics() noexcept {
+template <typename VectorType>
+void test_variables_prefix_math() noexcept {
+  using Prefix0VariablesType = Variables<
+      tmpl::list<VariablesTestTags_detail::PrefixTag0<
```
`add_prefix_to`

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -510,69 +510,128 @@ void test_variables_prefix_semantics() noexcept {
+  check_vectors(Prefix3VariablesType{number_of_grid_points,
+                                     value_in_variables / random_vals.at(0)},
+                prefix_vars0 / random_vals.at(0));
+  check_vectors(
```
explain why we might want to ever add things of different prefixes. @nilsdeppe @wthrowe @kidder Why do we want to do this?

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -510,69 +510,128 @@ void test_variables_prefix_semantics() noexcept {
+  check_vectors(Prefix0VariablesType{number_of_grid_points, expected},
+                test_assignment);
+  test_assignment -=
+      Prefix1VariablesType{number_of_grid_points, random_vals.at(0)};
```
`at(...)` ->`[...]` and everwhere else array is used.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -657,88 +657,135 @@ void test_variables_serialization() noexcept {
+  // Test assigning a single tag to a Variables with two tags, either from a
+  // Variables or a TaggedTuple
+  const auto first_test = [&number_of_grid_points, &values_in_variables ](
+      const auto& vars_subset0, const auto& vars_subset1) noexcept {
```
take more arguments to avoid assumptions in the lambda

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -657,88 +657,135 @@ void test_variables_serialization() noexcept {
+
+  // Test assigning a single tag to a Variables with two tags, either from a
+  // Variables or a TaggedTuple
+  const auto first_test = [&number_of_grid_points, &values_in_variables ](
```
give lambdas better names.

moxcodes commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -858,44 +858,82 @@ void test_variables_slice() noexcept {
+  // three vectors in the tensor. The slice1_vals should have been added to
+  // entries 2 and 6 in each vector.
+  auto expected = Tensor{
+      {{{orig_vals.at(0).at(0), orig_vals.at(0).at(1),
```
manually format this for readability. and use `// clang-format off` and `// clang-format on`

moxcodes reviewed (commented):
group code review with @fmahebert. @markscheel, @geoffrey4444 , and Rob

moxcodes commented:
fixed, rebased, and squashed.
Thanks for the reviews!

nilsdeppe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -547,68 +794,158 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Variables.SliceVariables",
+  // three vectors in the tensor. The slice1_vals should have been added to
+  // entries 2 and 6 in each vector.
+  // clang-format off
+  const Tensor expected = Tensor{
```
You can remove the ` = Tensor` and just do `const Tensor expected{` if you want (I'm fine with either, just an FYI. clang-tidy might care, not sure)

nilsdeppe reviewed (commented):
Okay, I'm happy with what you've done :)

@kidder @wthrowe could you please look at this when you get a chance? There are currently several separate commits to make it less overwhelming to review, but I think it should all go in as one commit. Please comment on whether or not you agree. Thanks!

wthrowe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -156,3 +157,16 @@ SPECTRE_ALWAYS_INLINE bool operator!=(const T& lhs,
+  template <typename ValType>
+  static SPECTRE_ALWAYS_INLINE SpinWeighted<R, S> apply(
+      const T& /*input*/, ValType value) noexcept {
+    return SpinWeighted<R, S>{make_with_value<R>(value, value)};
```
Shouldn't this pass `input` through to the inner `make_with_value`?

wthrowe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -166,364 +224,553 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Variables", "[DataStructures][Unit]") {
 #ifdef __clang__
 #pragma GCC diagnostic pop
 #endif  // defined(__clang__)
   // clang-tidy: false positive 'x' used after it was moved
```
`x` -> `move_to`

wthrowe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -176,23 +176,62 @@ ReturnType make_with_random_values(
     const gsl::not_null<RandomNumberDistribution*> distribution) noexcept {
-  T result{std::numeric_limits<tt::get_fundamental_type_t<T>>::signaling_NaN()};
+  T result{};
+  // T is a container type, so we fill with NaNs first
```
Why are we doing this?  We overwrite all the values on the following line.

moxcodes replied:
It's really only useful to ensure that all values are truly overwritten. If that's trustworthy enough, I'm happy to remove all of the initializations for these convenience functions.

wthrowe replied:
`fill_with_random_values` is a public interface, so I feel like not trusting it here (in the same file where it is defined) is strange because we do trust it everywhere else.  On the other hand, this is a testing utility, so some amount of redundant checking is fine.  I'll leave this one up to you or anyone else who has a preference.

nilsdeppe replied:
Yea, we should trust fill_with_random_values. Plus I don't think it actually fills anything with `NaN` because you defaulted constructed the `T`

moxcodes replied:
I'll remove the initializations, then. In these cases where `used_for_size` is not passed in, the container type is anticipated to be fixed-size type like complex or array, so default construction should provide an object with nonzero size to fill with random values.

wthrowe commented on src/Utilities/MakeWithValue.hpp:
```diff
@@ -41,17 +43,59 @@ namespace MakeWithValueImpls {
+struct MakeWithValueImpl<std::complex<R>, T> {
+  static SPECTRE_ALWAYS_INLINE std::complex<R> apply(const T& /* input */,
+                                                     const R value) noexcept {
+    return std::complex<R>(value, value);
```
This looks wrong, and all of these new impls need tests.

moxcodes replied:
I'm happy to add tests. Is the trouble here that it would be expected to assign just the real part of the complex? The motivation for this is to have a utility that can fill all entries of a multi-field type like complex with singaling_NaN during initialization.
If this isn't fitting for `make_with_value`, I can make a separate utility to `make_signaling_NaN` to  fit my use case.

wthrowe replied:
There are places in the code (`determinant_and_inverse`, at least) that assume `make_with_value<T>(..., 1.0)` is the multiplicative identity (for `T` where that makes sense).  A separate utility seems better.

nilsdeppe replied:
Yea... If I created it with `1` I would expect `1 + 0 i` not `1 + i`.

wthrowe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -166,364 +224,553 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Variables", "[DataStructures][Unit]") {
       expected_vars_sliced_in_y(x_extents * z_extents, 0.),
       expected_vars_sliced_in_z(x_extents * y_extents, 0.);
-  const size_t x_offset = 1, y_offset = 2, z_offset = 1;
+  const size_t x_offset = sdist(gen) % (x_extents - 1);
```
Why "- 1"?

wthrowe commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -547,68 +794,158 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Variables.SliceVariables",
+         orig_vals[0].at(3),
+         orig_vals[0].at(4) + slice0_vals[0].at(0),
+         orig_vals[0].at(5) + slice0_vals[0].at(1),
+         orig_vals[0].at(6) + slice1_vals[0].at(1) + slice0_vals[0].at(2),
```
The slice terms are in the opposite order on this line than the others.

nilsdeppe commented on src/Utilities/MakeNaN.hpp:
```diff
@@ -0,0 +1,40 @@
+/// \brief Returns an appropriate signaling NaN for fundamantal or multi-field
+/// types.
+template <typename T>
+T make_signaling_NaN() {
```
Just make two overloaded functions and remove the requires. The compiler error will be reasonable enough and shouldn't be encountered often, if ever. I don't see a reason to pay for SFINAE and all the extra code in this case.
```cpp
template <typename T>
T make_signaling_NaN(const T& /*meta*/) noexcept {
  return std::numeric_limits<T>::signaling_NaN();
}

template <typename T>
std::complex<T> make_signaling_NaN(const std::complex<T>& /*meta*/) noexcept {
  return {std::numeric_limits<T>::signaling_NaN(), std::numeric_limits<T>::signaling_NaN());
}
```

moxcodes replied:
This seems fine, but I would prefer the type to be an explicit template argument rather than passed in the argument, as that acts more similarly to the `std::numeric_limits`.

nilsdeppe replied:
`numeric_limits` is a struct with static functions, this already doesn't behave like `numeric_limits`. You could have a generic:
```cpp
template <typename T>
T make_signaling_NaN() noexcept {
  return make_signaling_NaN(static_cast<T>(0));
}
```
I think that'll do what you want without all the overhead of having to do memoization and SFINAE on top of pattern matching.

nilsdeppe commented on src/Utilities/MakeNaN.hpp:
```diff
@@ -0,0 +1,40 @@
+// Distributed under the MIT License.
```
The header name I think would be clearest if it were `MakeSignalingNan.hpp`. I don't think we'll ever have interest in a non-signaling NaN, and if we do we can rename it then.

nilsdeppe replied:
Could you please also ultimately add this utility in a separate commit before the Variables stuff?

nilsdeppe commented on tests/Unit/Utilities/Test_MakeNaN.cpp:
```diff
@@ -0,0 +1,59 @@
+
+namespace {
+
+// according to IEEE standards, NaNs must have all 1's in the exponent
```
Why not just use `std::isnan`? Does that not work? If you get FPE signals you can turn those off at the start of the test and back on later. I think that's clearer.

moxcodes replied:
Are you thinking of `std::signal`, or is there a more modern way of handling signals I'm not familiar with? because I'm finding that doesn't act quite as one would anticipate for FPE. It seems one cannot reliably ignore a SIGFPE with a do-nothing function in std::signal:
(from [cppreference](https://en.cppreference.com/w/cpp/utility/program/signal)):
"If the user defined function returns when handling SIGFPE, SIGILL, SIGSEGV or any other implementation-defined signal specifying a computational exception, the behavior is undefined."

nilsdeppe replied:
Sorry, if you include `ErrorHandling/FloatingPointExceptions.hpp` you can call `disable_floating_point_exceptions()` at the start of the test and `enable_floating_point_exceptions()` at the end. Then you should be able to just using `std::isnan`. I think all we care about is that they're set to `NaN`, not really that we can test for the signal.

moxcodes replied:
Seems close enough to me. :+1:

nilsdeppe commented on src/DataStructures/DataVector.hpp:
```diff
@@ -481,12 +481,11 @@ namespace MakeWithValueImpls {
-};
+SPECTRE_ALWAYS_INLINE DataVector
+MakeWithValueImpl<DataVector, DataVector>::apply(const DataVector& input,
+                                                 const double value) {
```
You've lost a lot of `noexcept`s in this commit

moxcodes replied:
This just reverts the `MakeWithValueImpl` to the state in develop, which had many missing `noexcept`s. With the MakeNaN utility, this PR no longer touches `MakeWithValue` at all

nilsdeppe replied:
Okay, I'll revisit this at the end when your additions are clearer :)

moxcodes replied:
would it be preferable to squash this part so it vanishes from the fixup diffs?

nilsdeppe replied:
I wouldn't object to that, personally. I would find it easier to review, though I of course can't speak for other people :)

nilsdeppe reviewed (changes requested):
also, please format your code before committing.

moxcodes committed:
Add utilitiy for making signaling NaNs

- forwards to `numeric_limits` for many cases
- fills both entries of a complex with signaling NaN

moxcodes committed:
Template Variables on contained type

- Tensor is templated on type, so Variables can easily be templated on the type
  to be stored in the Tensor.
- Necessary for usage of extensions such as ComplexDataVector

nilsdeppe commented on src/Utilities/MakeSignalingNan.hpp:
```diff
@@ -0,0 +1,27 @@
+// @{
+/// \ingroup UtilitiesGroup
+/// \brief Returns an appropriate signaling NaN for fundamantal or multi-field
+/// types.
```
Maybe add "(such as `std::complex`)" to be very clear :)

nilsdeppe reviewed (commented):
Okay, I'm happy with the fixups. @wthrowe could you please check the fixups? @kidder please let us know if you want to review this or not. I know you're quite busy right now :)

wthrowe commented:
Fixups look fine.

nilsdeppe reviewed (commented):
Okay, looks good to squash :)

nilsdeppe commented:
Woops, at least the fixups, I think we still need feedback from @kidder

moxcodes commented:
fixups squashed, but not the extra commits for all the tests.
Cheers!

kidder commented on tests/Unit/DataStructures/Test_Variables.cpp:
```diff
@@ -85,76 +104,135 @@ static_assert(
+            filled_variables.number_of_independent_components);
+  CHECK(number_of_grid_points == filled_variables.number_of_grid_points());
+  CHECK(number_of_grid_points * 5 == filled_variables.size());
+  CHECK(initial_fill_value == filled_variables.data()[0]);  // NOLINT
```
Shouldn't we test the whole variables?

moxcodes replied:
It seems like we probably should. I'll update this one as well as the several below tests which just check the first element.

kidder commented:
@wthrowe @nilsdeppe please look at the fixup commit from the one change I requested.  I am happy for everything to be squashed if @wthrowe and @nilsdeppe are happy with the fixup

nilsdeppe commented:
I'm happy with it :)

wthrowe commented:
Squash.

moxcodes commented:
squashed :octopus: 
Cheers!

nilsdeppe reviewed (approved):
I've built each commit individually, ran clang-tidy and iwyu on each commit, and ran tests. 

@kidder @wthrowe could you please give feedback? Should be ready :)
